File: nuwa-script/implementations/python/nuwa/parser.py
import ply.yacc as yacc

# Get the token map from the lexer. This is required.
from .lexer import tokens, lexer
from .ast import (
    Script, Statement, LetStatement, CallStatement, IfStatement, ForStatement,
    Expression, Literal, Variable, BinaryOp, UnaryOp, FunctionCall,
    CallExpression, CalcExpression
)

# --- Operator Precedence ---
# Define operator precedence and associativity
precedence = (
    ('left', 'OR'),
    ('left', 'AND'),
    ('right', 'NOT'),
    ('nonassoc', 'EQ', 'NE', 'LT', 'LE', 'GT', 'GE'), # Nonassociative: a < b < c is illegal
    # Add precedence for other potential operators like +, -, *, / if CALC is expanded
)

# --- Grammar Rules ---

# Top-level rule: script is a list of statements
def p_script(p):
    '''script : statements'''
    p[0] = Script(statements=p[1])

def p_statements(p):
    '''statements : statements statement
                  | statement'''
    if len(p) == 3:
        p[0] = p[1] + [p[2]] # Append statement to list
    else:
        p[0] = [p[1]] # Start a new list

# Rule for individual statement types
def p_statement(p):
    '''statement : let_statement
                 | call_statement
                 | if_statement
                 | for_statement'''
    p[0] = p[1]

# LET statement: LET ID = expression
def p_let_statement(p):
    '''let_statement : LET ID ASSIGN expression'''
    p[0] = LetStatement(variable_name=p[2], value=p[4])

# CALL statement: CALL ID { arguments }
def p_call_statement(p):
    '''call_statement : CALL ID LBRACE arguments RBRACE'''
    p[0] = CallStatement(tool_name=p[2], arguments=p[4])

# IF statement: IF expression THEN statements [ELSE statements] END
def p_if_statement(p):
    '''if_statement : IF expression THEN statements END
                    | IF expression THEN statements ELSE statements END'''
    if len(p) == 6: # IF ... THEN ... END
        p[0] = IfStatement(condition=p[2], then_block=p[4], else_block=None)
    else: # IF ... THEN ... ELSE ... END
        p[0] = IfStatement(condition=p[2], then_block=p[4], else_block=p[6])

# FOR statement: FOR ID IN expression DO statements END
def p_for_statement(p):
    '''for_statement : FOR ID IN expression DO statements END'''
    p[0] = ForStatement(iterator_variable=p[2], iterable=p[4], loop_block=p[6])

# --- Arguments for CALL and CALC ---

# Arguments: key: value, key: value, ...
def p_arguments(p):
    '''arguments : argument_list
                 | empty'''
    p[0] = p[1] if p[1] is not None else {}

def p_argument_list(p):
    '''argument_list : argument_list COMMA argument
                     | argument'''
    if len(p) == 4:
        p[1][p[3][0]] = p[3][1] # Add key-value pair to dict
        p[0] = p[1]
    else:
        p[0] = {p[1][0]: p[1][1]} # Start a new dict

def p_argument(p):
    '''argument : ID COLON expression'''
    p[0] = (p[1], p[3]) # Return tuple (key, value_expr)

def p_empty(p):
    '''empty :'''
    p[0] = None

# --- Expressions ---

# Base expression rule
def p_expression(p):
    '''expression : binary_op
                  | unary_op
                  | function_call
                  | call_expression
                  | calc_expression
                  | literal
                  | variable
                  | LPAREN expression RPAREN'''
    if len(p) == 2:
        p[0] = p[1] # Simple expression
    else: # Parenthesized expression
        p[0] = p[2]

# Binary operations: expression OP expression
def p_binary_op(p):
    '''binary_op : expression EQ expression
                 | expression NE expression
                 | expression LT expression
                 | expression LE expression
                 | expression GT expression
                 | expression GE expression
                 | expression AND expression
                 | expression OR expression
                 | expression DOT ID ''' # Basic member access

    if p[2] == '==': p[0] = BinaryOp(operator='==', left=p[1], right=p[3])
    elif p[2] == '!=': p[0] = BinaryOp(operator='!=', left=p[1], right=p[3])
    elif p[2] == '<': p[0] = BinaryOp(operator='<', left=p[1], right=p[3])
    elif p[2] == '<=': p[0] = BinaryOp(operator='<=', left=p[1], right=p[3])
    elif p[2] == '>': p[0] = BinaryOp(operator='>', left=p[1], right=p[3])
    elif p[2] == '>=': p[0] = BinaryOp(operator='>=', left=p[1], right=p[3])
    elif p[2].upper() == 'AND': p[0] = BinaryOp(operator='AND', left=p[1], right=p[3])
    elif p[2].upper() == 'OR': p[0] = BinaryOp(operator='OR', left=p[1], right=p[3])
    elif p[2] == '.':
        # This is a simplified member access, assuming left is Variable or similar
        # Might need a dedicated MemberAccess node in AST for proper handling
        if isinstance(p[1], Variable):
             # Represent as Variable with dotted name for now
             p[0] = Variable(name=f"{p[1].name}.{p[3]}")
        else:
             # Handle error or more complex cases later
             print(f"Syntax Error: Cannot access member '{p[3]}' of non-variable {p[1]}")
             p[0] = None # Or raise specific parse error

# Unary operations: NOT expression
def p_unary_op(p):
    '''unary_op : NOT expression'''
    p[0] = UnaryOp(operator='NOT', operand=p[2])

# Built-in function call: NOW()
def p_function_call(p):
    '''function_call : NOW LPAREN RPAREN'''
    p[0] = FunctionCall(function_name='NOW')
    # Add rules for functions with arguments if needed later

# CALL used as expression: CALL ID { arguments }
def p_call_expression(p):
    '''call_expression : CALL ID LBRACE arguments RBRACE'''
    p[0] = CallExpression(tool_name=p[2], arguments=p[4])

# CALC expression: CALC { formula: STRING, vars: { arguments } }
def p_calc_expression(p):
    '''calc_expression : CALC LBRACE calc_args RBRACE'''
    p[0] = CalcExpression(formula=p[3]['formula'], variables=p[3]['vars'])

def p_calc_args(p):
    '''calc_args : ID COLON STRING COMMA ID COLON LBRACE arguments RBRACE'''
    formula_key = p[1]
    vars_key = p[5]
    if formula_key != 'formula':
        raise SyntaxError(f"Expected 'formula' key in CALC block, got '{formula_key}' at line {p.lineno(1)}")
    if vars_key != 'vars':
        raise SyntaxError(f"Expected 'vars' key in CALC block, got '{vars_key}' at line {p.lineno(5)}")

    p[0] = {'formula': p[3], 'vars': p[8]}

# Literal values
def p_literal(p):
    '''literal : NUMBER
               | STRING
               | BOOLEAN'''
               # | NULL # Add if needed
    p[0] = Literal(value=p[1])

# Variable access
def p_variable(p):
    '''variable : ID'''
    p[0] = Variable(name=p[1])

# Error rule for syntax errors
def p_error(p):
    if p:
        print(f"Syntax error at token {p.type} ('{p.value}') on line {p.lineno}")
        # Potentially try to recover or just stop
    else:
        print("Syntax error at EOF")

# Build the parser
parser = yacc.yacc()

# --- Helper function for testing ---
def parse_script(script_content: str) -> Script | None:
    """Parses a NuwaScript string into an AST."""
    # Reset lexer state for each parse
    lexer.lineno = 1
    result = parser.parse(script_content, lexer=lexer)
    return result

if __name__ == '__main__':
    # Example usage for testing
    test_script = """
    LET price = CALL get_price { token: "BTC" }
    LET threshold = 70000
    IF price < threshold AND price > 0 THEN
      LET action = "buy"
      CALL swap { from_token: "USDT", to_token: "BTC", amount: 100 }
    ELSE
      LET action = "wait"
      CALL reply { message: "Price not in range" }
    END

    FOR nft IN listed_nfts DO
        IF nft.rarity > 90 THEN
            CALL buy_nft { id: nft.id, price: nft.price }
        END
    END

    LET current_time = NOW()
    LET estimated_value = CALC {
        formula: "price * 1.1",
        vars: { price: 500 }
     }
    """
    ast = parse_script(test_script)
    if ast:
        import json
        # Crude way to print AST using dataclasses asdict helper
        from dataclasses import asdict

        def default_serializer(obj):
             if dataclasses.is_dataclass(obj):
                 return asdict(obj)
             raise TypeError(f"Object of type {type(obj).__name__} is not JSON serializable")

        # Print a JSON-like representation
        # This will fail if nodes contain non-serializable types directly
        # print(json.dumps(ast, default=default_serializer, indent=2))
        # Safer: just print the structure
        print(ast)


    # Test error case
    error_script = "LET x = 5 IF x > THEN CALL foo END"
    print("\nTesting error case:")
    parse_script(error_script)

File: nuwa-script/implementations/typescript/src/parser.ts
import { Token, TokenType, tokenize } from './lexer';
import * as AST from './ast';
import { NuwaValue } from './values'; // For literal parsing

class ParserError extends Error {
    constructor(message: string, public token?: Token) {
        const position = token ? ` at ${token.line}:${token.column}` : '';
        super(`${message}${position}`);
        this.name = 'ParserError';
        Object.setPrototypeOf(this, ParserError.prototype);
    }
}

export class Parser {
    private tokens: Token[] = [];
    private current: number = 0; // Index of the next token to consume

    constructor(tokens: Token[]) {
        // Filter out whitespace and comments before parsing
        this.tokens = tokens.filter(token =>
            token.type !== TokenType.WHITESPACE && token.type !== TokenType.COMMENT
        );
        this.current = 0;
    }

    /**
     * Parses the token stream into a Script AST node.
     */
    public parse(): AST.Script {
        const statements: AST.Statement[] = [];
        while (!this.isAtEnd()) {
            statements.push(this.parseStatement());
        }
        return { kind: 'Script', statements };
    }

    // --- Statement Parsing ---

    private parseStatement(): AST.Statement {
        switch (this.peek().type) {
            case TokenType.LET:
                return this.parseLetStatement();
            case TokenType.CALL:
                // Check if it's CALL used as statement or expression context (not really possible at statement level)
                return this.parseCallStatement();
             case TokenType.PRINT:
                return this.parsePrintStatement();
            case TokenType.IF:
                return this.parseIfStatement();
            case TokenType.FOR:
                return this.parseForStatement();
            default:
                throw new ParserError(`Unexpected token type '${this.peek().type}' at start of statement`, this.peek());
                // Or potentially try parsing an expression statement if the language supported it
        }
    }

    private parseLetStatement(): AST.LetStatement {
        this.consume(TokenType.LET, "Expected 'LET' keyword.");
        const nameToken = this.consume(TokenType.IDENTIFIER, "Expected variable name after 'LET'.");
        this.consume(TokenType.ASSIGN, "Expected '=' after variable name.");
        const value = this.parseExpression();
        // Optional: Expect END or newline? For now, assume simple structure.
        return { kind: 'LetStatement', variableName: nameToken.value, value };
    }

     private parseCallStatement(): AST.CallStatement {
        this.consume(TokenType.CALL, "Expected 'CALL' keyword.");
        const nameToken = this.consume(TokenType.IDENTIFIER, "Expected tool name after 'CALL'.");
        const args = this.parseArguments();
        return { kind: 'CallStatement', toolName: nameToken.value, arguments: args };
    }

    private parsePrintStatement(): AST.PrintStatement {
        this.consume(TokenType.PRINT, "Expected 'PRINT' keyword.");
        this.consume(TokenType.LPAREN, "Expected '(' after 'PRINT'.");
        const value = this.parseExpression();
        this.consume(TokenType.RPAREN, "Expected ')' after PRINT expression.");
        return { kind: 'PrintStatement', value };
    }

     private parseIfStatement(): AST.IfStatement {
        this.consume(TokenType.IF, "Expected 'IF' keyword.");
        const condition = this.parseExpression();
        this.consume(TokenType.THEN, "Expected 'THEN' after IF condition.");

        const thenBlock: AST.Statement[] = [];
        while (!this.check(TokenType.ELSE) && !this.check(TokenType.END) && !this.isAtEnd()) {
            thenBlock.push(this.parseStatement());
        }

        let elseBlock: AST.Statement[] | undefined = undefined;
        if (this.match(TokenType.ELSE)) {
            elseBlock = [];
            while (!this.check(TokenType.END) && !this.isAtEnd()) {
                elseBlock.push(this.parseStatement());
            }
        }

        this.consume(TokenType.END, "Expected 'END' to close 'IF' statement.");
        return { kind: 'IfStatement', condition, thenBlock, elseBlock };
    }

     private parseForStatement(): AST.ForStatement {
        this.consume(TokenType.FOR, "Expected 'FOR' keyword.");
        const iteratorToken = this.consume(TokenType.IDENTIFIER, "Expected iterator variable name after 'FOR'.");
        this.consume(TokenType.IN, "Expected 'IN' after iterator variable.");
        const iterable = this.parseExpression();
        this.consume(TokenType.DO, "Expected 'DO' after iterable expression.");

        const loopBlock: AST.Statement[] = [];
        while (!this.check(TokenType.END) && !this.isAtEnd()) {
            loopBlock.push(this.parseStatement());
        }

        this.consume(TokenType.END, "Expected 'END' to close 'FOR' statement.");
        return { kind: 'ForStatement', iteratorVariable: iteratorToken.value, iterable, loopBlock };
    }

    // --- Argument Parsing for CALL ---

    private parseArguments(): Record<string, AST.Expression> {
        this.consume(TokenType.LBRACE, "Expected '{' to start arguments block.");
        const args: Record<string, AST.Expression> = {};
        if (!this.check(TokenType.RBRACE)) { // Check if block is not empty
            do {
                const keyToken = this.consume(TokenType.IDENTIFIER, "Expected argument name.");
                this.consume(TokenType.COLON, "Expected ':' after argument name.");
                const value = this.parseExpression();
                args[keyToken.value] = value;
            } while (this.match(TokenType.COMMA)); // Continue if comma is found
        }
        this.consume(TokenType.RBRACE, "Expected '}' to end arguments block.");
        return args;
    }


    // --- Expression Parsing (Recursive Descent with Precedence) ---

    private parseExpression(): AST.Expression {
        return this.parseLogicalOr(); // Start with lowest precedence (OR)
    }

    private parseLogicalOr(): AST.Expression {
        let expr = this.parseLogicalAnd();
        while (this.match(TokenType.OR)) {
            const operator = this.previous().type as AST.BinaryOperator; // Should be 'OR'
            const right = this.parseLogicalAnd();
            expr = { kind: 'BinaryOpExpr', operator, left: expr, right };
        }
        return expr;
    }

    private parseLogicalAnd(): AST.Expression {
        let expr = this.parseEquality();
        while (this.match(TokenType.AND)) {
            const operator = this.previous().type as AST.BinaryOperator; // Should be 'AND'
            const right = this.parseEquality();
            expr = { kind: 'BinaryOpExpr', operator, left: expr, right };
        }
        return expr;
    }

    private parseEquality(): AST.Expression {
        let expr = this.parseComparison();
        while (this.match(TokenType.EQ, TokenType.NE)) {
            const operator = this.previous().type as AST.BinaryOperator; // '==' or '!='
            const right = this.parseComparison();
            expr = { kind: 'BinaryOpExpr', operator, left: expr, right };
        }
        return expr;
    }

    private parseComparison(): AST.Expression {
        let expr = this.parseTerm(); // Parse higher precedence: +, -
        while (this.match(TokenType.GT, TokenType.GE, TokenType.LT, TokenType.LE)) {
             const operator = this.previous().type as AST.BinaryOperator; // '>', '>=', '<', '<='
            const right = this.parseTerm();
            expr = { kind: 'BinaryOpExpr', operator, left: expr, right };
        }
        return expr;
    }

    // Term handles Addition (+) and Subtraction (-)
    private parseTerm(): AST.Expression {
         let expr = this.parseFactor(); // Parse higher precedence first
         while (this.match(TokenType.PLUS, TokenType.MINUS)) {
            const operator = this.previous().type as AST.BinaryOperator; // '+' or '-'
            const right = this.parseFactor();
            expr = { kind: 'BinaryOpExpr', operator, left: expr, right };
         }
         return expr;
    }

    // Factor handles Multiplication (*) and Division (/)
    private parseFactor(): AST.Expression {
        // Factor now calls parseUnary, which handles postfix operations
        let expr = this.parseUnary();
        while (this.match(TokenType.STAR, TokenType.SLASH)) {
            const operator = this.previous().type as AST.BinaryOperator; // '*' or '/'
            // Note: The right operand should also potentially handle postfix ops
            const right = this.parseUnary();
            expr = { kind: 'BinaryOpExpr', operator, left: expr, right };
        }
        return expr;
    }

    private parseUnary(): AST.Expression {
        // Handle unary PLUS and MINUS (as well as NOT)
        if (this.match(TokenType.NOT, TokenType.PLUS, TokenType.MINUS)) {
            const operator = this.previous().type as AST.UnaryOperator;
            // Unary operators typically apply to the result of the next highest precedence level,
            // which could potentially be another unary operation (e.g., - - 1)
            // or a primary expression.
            // Let's have unary parse unary for right-associativity (or simply primary if simpler).
            const operand = this.parseUnary(); // Recursively parse operand
            return { kind: 'UnaryOpExpr', operator, operand };
        }
        // If no unary operator, parse the next level down
        // Instead of primary, call postfix which handles primary AND subsequent [] . ()
        return this.parsePostfix(); // CHANGED: Call parsePostfix here
    }

    // NEW METHOD: Handles postfix operations like array indexing [], member access ., and function calls ()
    private parsePostfix(): AST.Expression {
        // First, parse the primary expression (the base)
        let expr = this.parsePrimary();

        // Then, loop to check for postfix operators
        while (true) {
            if (this.match(TokenType.LBRACKET)) {
                // Array index operator '['
                const index = this.parseExpression(); // Parse the index expression
                this.consume(TokenType.RBRACKET, "Expected ']' after array index expression.");
                // Create the ArrayIndexExpression node, using the current expression as the object
                expr = { kind: 'ArrayIndexExpression', object: expr, index };
            } else if (this.match(TokenType.DOT)) { // Enable DOT handling
                // Member Access operator '.'
                const propertyToken = this.consume(TokenType.IDENTIFIER, "Expected property name after '.'.");
                expr = { kind: 'MemberAccessExpr', object: expr, property: propertyToken.value };
            }
            /* // Future Extension: Function Call () - Needs more sophisticated argument parsing integration
             else if (this.match(TokenType.LPAREN)) {
                 // Parse arguments...
                 this.consume(TokenType.RPAREN, "Expected ')' after arguments.");
             */
            else {
                // No more postfix operators, break the loop
                break;
            }
        }

        return expr;
    }

    private parsePrimary(): AST.Expression {
        // Literals
        if (this.match(TokenType.TRUE)) return { kind: 'LiteralExpr', value: true };
        if (this.match(TokenType.FALSE)) return { kind: 'LiteralExpr', value: false };
        if (this.match(TokenType.NULL)) return { kind: 'LiteralExpr', value: null };
        if (this.match(TokenType.NUMBER)) {
            return { kind: 'LiteralExpr', value: parseFloat(this.previous().value) };
        }
        if (this.match(TokenType.STRING)) {
             // Lexer already processed escapes and removed quotes via JSON.parse
            return { kind: 'LiteralExpr', value: this.previous().value };
        }

        // List Literal [ ... ]
        if (this.match(TokenType.LBRACKET)) {
            const elements: AST.Expression[] = [];
            // To handle potential ambiguity with index operator, we must *not* consume LBRACKET here
            // if it's followed by an expression and RBRACKET in the context of parsePostfix.
            // However, parsePrimary is called first. Let's assume list literals are distinct.
            // The `match` call above already consumed LBRACKET for the list literal case.
            if (!this.check(TokenType.RBRACKET)) { // Handle non-empty list
                do {
                    if (this.check(TokenType.RBRACKET)) break;
                    elements.push(this.parseExpression());
                } while (this.match(TokenType.COMMA));
            }
            this.consume(TokenType.RBRACKET, "Expected ']' after list elements.");

             // !!! Important Correction for List Literals !!!
            // The interpreter needs the evaluated values, not expressions, for a literal.
            // We need a specific AST node for List Literals, or handle evaluation here.
            // Let's introduce ListLiteralExpr to AST first.
            // For now, let's keep parsing elements as Expressions and create a generic
            // structure that the interpreter will handle. This is complex.

            // **Temporary simplification**: Create a LiteralExpr with a placeholder or
            // leave elements as expressions for the interpreter to resolve.
            // Let's create a dedicated ListLiteralExpr node in AST later if needed.
            // For now, let's assume the interpreter can handle evaluating expressions
            // inside a structure representing the list.
            // How to represent this in AST? Maybe just use LiteralExpr with an array value?
            // This is tricky because the *elements* are expressions, not values yet.
            //
            // Let's stick to the original plan of having elements as Expression[] and
            // create a dedicated node type if issues arise. The interpreter will evaluate them.
            // Let's define ListLiteralExpr in ast.ts NEXT.
            // *** TEMPORARY HACK for now: Return a LiteralExpr that the interpreter won't handle correctly ***
            // We need to add ListLiteralExpr to ast.ts and modify this code block.
            // For the current goal of ARRAY INDEXING, let's just throw an error for list literals for now.
             throw new ParserError("List literals [ ... ] are not fully implemented yet.", this.previous());

             /* // Ideal future code after adding ListLiteralExpr to ast.ts:
              return { kind: 'ListLiteralExpr', elements };
             */
        }

        // Variable (single identifier only now)
        if (this.match(TokenType.IDENTIFIER)) {
            // Now only handles simple 'variable', not 'variable.dotted.name'
            // Dotted names are handled by parsePostfix recognizing the DOT token.
            return { kind: 'VariableExpr', name: this.previous().value };
        }

        // Function Calls (NOW())
        if (this.match(TokenType.NOW)) {
            this.consume(TokenType.LPAREN, "Expected '(' after 'NOW'.");
            this.consume(TokenType.RPAREN, "Expected ')' after 'NOW()'.");
            return { kind: 'FunctionCallExpr', functionName: 'NOW' };
        }

        // Tool Calls used as expressions
        if (this.match(TokenType.CALL)) {
            // Need to backtrack slightly as we consumed CALL
            this.current--; // Go back to CALL token
            return this.parseToolCallExpression();
        }

        // Parenthesized expression
        if (this.match(TokenType.LPAREN)) {
            const expr = this.parseExpression();
            this.consume(TokenType.RPAREN, "Expected ')' after expression.");
            return expr;
        }

        // Error
        throw new ParserError(`Unexpected token type '${this.peek().type}' in expression`, this.peek());
    }

     // NEW HELPER for CALL used as an expression
    private parseToolCallExpression(): AST.ToolCallExpr {
         this.consume(TokenType.CALL, "Expected 'CALL' keyword."); // Already checked, but good practice
         const nameToken = this.consume(TokenType.IDENTIFIER, "Expected tool name after 'CALL'.");
         const args = this.parseArguments(); // Reuse argument parsing
         return { kind: 'ToolCallExpr', toolName: nameToken.value, arguments: args };
    }

    // --- Parser Helpers ---

    // Checks if the current token is of the given type(s) without consuming it.
    private check(...types: TokenType[]): boolean {
        if (this.isAtEnd()) return false;
        return types.includes(this.peek().type);
    }

    // Consumes the current token if it matches one of the given types.
    private match(...types: TokenType[]): boolean {
        if (this.check(...types)) {
            this.advance();
            return true;
        }
        return false;
    }

    // Consumes the current token if it's of the expected type, otherwise throws an error.
    private consume(type: TokenType, errorMessage: string): Token {
        if (this.check(type)) {
            return this.advance();
        }
        throw new ParserError(errorMessage, this.peek());
    }

    // Moves to the next token and returns the previous one.
    private advance(): Token {
        if (!this.isAtEnd()) {
            this.current++;
        }
        return this.previous();
    }

    // Checks if we've run out of tokens (ignoring EOF).
    private isAtEnd(): boolean {
        return this.peek().type === TokenType.EOF;
    }

    // Returns the current token without consuming it.
    private peek(): Token {
        return this.tokens[this.current]!;
    }

    // Returns the most recently consumed token.
    private previous(): Token {
        return this.tokens[this.current - 1]!;
    }
}

/**
 * Convenience function to parse NuwaScript source code directly.
 * @param sourceCode The NuwaScript code string.
 * @returns The parsed Script AST node.
 * @throws LexerError or ParserError on failure.
 */
export function parse(sourceCode: string): AST.Script {
    const tokens = tokenize(sourceCode);
    const parser = new Parser(tokens);
    return parser.parse();
}
